### PR TITLE
[6/10] Code generation for fully connected layer via CMSIS-NN

### PR DESCRIPTION
**Pull request by ashutosh-arm**
Support for fully connected layer via CMSIS-NN.

Here is the tracking issue: #8646

### CODE REVIEW

**Comment by ashutosh-arm** [comment]
cc: @Mousius @manupa-arm @leandron @areusch for code reviews.

File: src/relay/backend/contrib/cmsisnn/relay_to_tir.cc [hunk 1]
```diff
@@ -234,6 +222,114 @@ class RelayToTIRVisitor : public MixedModeMutator {
...
+    std::string context_buffer_name = "NULL";
+    tvm::Array<PrimExpr> context_buffer_args = {tir::StringImm(context_buffer_name),
+                                                ToArg(context_buffer_size)};
+
+    scalar_args = tvm::runtime::Concat(context_buffer_args, scalar_args);
+    scalar_args = tvm::runtime::Concat(scalar_args, cmsisnn_input_shape);
+    scalar_args = tvm::runtime::Concat(scalar_args, cmsisnn_filter_shape);
+    scalar_args = tvm::runtime::Concat(scalar_args, bias_shape);
+    scalar_args = tvm::runtime::Concat(scalar_args, cmsisnn_output_shape);
+    call_ext_args = tvm::runtime::Concat(call_ext_args, scalar_args);
```
Comment (manupa-arm) [inline src/relay/backend/contrib/cmsisnn/relay_to_tir.cc:322]: nit : Would it be possible to keep the definitions of sub-catergories that get concatenated closer to here ?  So it is easy to follow ?

**Review by manupa-arm — CHANGES_REQUESTED** [review]
Broadly looks good! Few stylistic comments!

File: src/relay/backend/contrib/cmsisnn/relay_to_tir.cc [hunk 2]
```diff
@@ -234,6 +222,114 @@ class RelayToTIRVisitor : public MixedModeMutator {
...
+        static_cast<double>(input_scale) / static_cast<double>(output_scale);
+    auto mult_shift_pair = tvm::relay::qnn::GetFixedPointMultiplierShift(quantized_multiplier);
+    int32_t multiplier = std::get<0>(mult_shift_pair);
+    int32_t shift = std::get<1>(mult_shift_pair);
+
+    tvm::Array<PrimExpr> scalar_args = {
+        ToArg(input_offset), ToArg(filter_offset), ToArg(output_offset), ToArg(clip_min),
+        ToArg(clip_max),     ToArg(multiplier),    ToArg(shift)};
+
+    // cmsis_nn_dims *input_dims
```
[the remark this answers is not in the document]
Comment (ashutosh-arm) [reply src/relay/backend/contrib/cmsisnn/relay_to_tir.cc:290]: They are the names of the CMSIS-NN structs. Someone developing / maintaining the code can immediately look them up into the CMSIS-NN repo.

[the remark this answers is not in the document]
Comment (ashutosh-arm) [reply src/relay/backend/contrib/cmsisnn/relay_to_tir.cc:290]: But I do see the problem as just this comment may not suffice.

File: src/relay/backend/contrib/cmsisnn/relay_to_tir.cc (hunk 1 above)
Comment (ashutosh-arm) [reply src/relay/backend/contrib/cmsisnn/relay_to_tir.cc:322]: I didn't get this. Which ones should be close together and are not?

Comment (ashutosh-arm) [reply src/relay/backend/contrib/cmsisnn/relay_to_tir.cc:322]: I see what you mean

Comment (ashutosh-arm) [reply src/relay/backend/contrib/cmsisnn/relay_to_tir.cc:322]: Can't do because of a cross dependency in this case.

**Pull request closed without merging**